lib/internal/streams/duplexify.js

```diff
@@ -0,0 +1,286 @@
+'use strict';
+
+const {
+  isReadable,
+  isWritable,
+  isIterable,
+  isNodeStream,
+  isReadableNodeStream,
+  isWritableNodeStream,
+  isDuplexNodeStream,
+} = require('internal/streams/utils');
+const eos = require('internal/streams/end-of-stream');
+const {
+  AbortError,
+  codes: {
+    ERR_INVALID_ARG_TYPE,
+    ERR_INVALID_RETURN_VALUE,
+  },
+} = require('internal/errors');
+const { destroyer } = require('internal/streams/destroy');
+const Duplex = require('internal/streams/duplex');
+const Readable = require('internal/streams/readable');
+const Writable = require('internal/streams/writable');
+const { createDeferredPromise } = require('internal/util');
+const from = require('internal/streams/from');
+
+const {
+  isBlob,
+} = require('internal/blob');
+
+const {
+  isBrandCheck,
+} = require('internal/webstreams/util');
+
+const isReadableStream =
+  isBrandCheck('ReadableStream');
+const isWritableStream =
+  isBrandCheck('WritableStream');
+
+const {
+  PromiseResolve
+} = primordials;
+
+module.exports = function duplexify(body, name) {
+  if (isDuplexNodeStream(body)) {
+    return body;
+  } else if (isReadableNodeStream(body)) {
+    return _duplexify({ readable: body });
+  } else if (isWritableNodeStream(body)) {
+    return _duplexify({ writable: body });
+  } else if (isNodeStream(body)) {
+    return _duplexify({ writable: false, readable: false });
+  } else if (isReadableStream(body)) {
+    return _duplexify({ readable: Readable.fromWeb(body) });
+  } else if (isWritableStream(body)) {
+    return _duplexify({ writable: Writable.fromWeb(body) });
+  } else if (typeof body === 'function') {
+    const { value, write, final } = fromAsyncGen(body);
+
+    if (isIterable(value)) {
+      return from(Duplex, value, {
+        objectMode: true,
+        highWaterMark: 1,
+        write,
+        final
+      });
+    }
+
+    if (typeof value?.then === 'function') {
+      let d;
+
+      const promise = PromiseResolve(value)
+        .then((val) => {
+          if (val != null) {
+            throw new ERR_INVALID_RETURN_VALUE('nully', 'body', val);
+          }
+        })
+        .catch((err) => {
+          destroyer(d, err);
+        });
```

Comment:
We should support Promises/A+ spec: https://github.com/nodejs/node/blob/5ce015ec72be98f064041d1bf5c3527a89c276cc/lib/events.js#L223-L235
```suggestion
    const { then } = value;
    if (typeof then === 'function') {
      let d;

      const promise = FunctionPrototypeCall(then, value,
        (val) => {
          if (val != null) {
            throw new ERR_INVALID_RETURN_VALUE('nully', 'body', val);
          }
        },
        (err) => {
          destroyer(d, err);
        }));
```
I see that this wasn't introduced by this PR, but it's still worth fixing IMO.